proposals/0046-actions.md

```diff
@@ -0,0 +1,637 @@
+        // handle result
+      }}
+    >
+      <input type="hidden" name="postId" value={postId} />
+      <label for="author">Author</label>
+      <input id="author" type="text" name="author" />
+      <textarea rows={10} name="body"></textarea>
+      <button type="submit">Post</button>
+    </form>
+  );
+}
+```
+
+### Add progressive enhancement with fallbacks
+
+> Note: You will need server-rendering for progressive fallbacks. Be sure to [opt out of prerendering](https://docs.astro.build/en/guides/server-side-rendering/#opting-out-of-pre-rendering-in-hybrid-mode) on the page containing a progressively enhanced form.
+
+Actions work well when client JavaScript is enabled. However, you may see unexpected behavior when a form is submitted _before_ your component's JavaScript loads. This is common on slow internet connections or older devices. To offer a fallback experience, you can add a progressive fallback to your form.
+
+To add a fallback, add the `method="POST"` property to your `<form>` element. Then, call the `getNameProps()` function from `astro:actions` with the action you want to use (ex. `getNameProps(actions.comment)`). Spread the result onto an `input` within your form to apply metadata for Astro to handle the request:
+
+```tsx
+import { actions, getNameProps } from 'astro:actions';
+
+// src/components/Comment.jsx
+export function Comment({ postId }: { postId: string }) {
+	return (
+		<form method="POST" onSubmit={...}>
+			<input {...getNameProps(actions.comment)} />
+			{/* result:
+			<input
+				type="hidden"
+				name="_astroAction"
+				value="/_actions/comment" />
+			*/}
+		</form>
+	)
+}
+```
+
+Implementation: The `getNameProps()` function returns hidden input attributes to tell Astro which action should handle the request. Using middleware, Astro will intercept incoming requests with a form data `Content-type`. The form data is parsed to check for the `_astroAction` field and call the appropriate action.
+
+### Handle an action result on the server
+
+When using a progressive fallback, you can get the result of an action from your Astro frontmatter. Call `Astro.getActionResult()` with the action you want (ex. `Astro.getActionResult(actions.comment)`). This will return type-safe data when a matching POST request is received and `undefined` otherwise.
+
+```astro
+---
+import { actions } from 'astro:actions';
+
+const comment = Astro.getActionResult(actions.comment);
+---
+
+{comment && (
+	<article class="new-comment">
+		{/* ... */}
+	</article>
+)}
+```
+
+## Handle errors
+
+Actions can raise exceptions for any number of reasons: an invalid input, an unauthorized request, a missing resource, etc. By default, these errors will `throw` wherever your action is called.
+
+To handle errors in your code without using a `try / catch`, you can chain the `.safe()` function onto your existing actions. This will change the return type to either a `data` object with the result, or an `error` object with the exception.
```

Comment:
Reading this section I can't help but think that I always want to use `safe()`. APIs can fail, especially when validation is involved. Is there a strong use-case for every using the non-safe version of this?

Reply:
That's fair! Though with JSON APIs, I've found it's rare that I want to handle validation exceptions since I trust the type inference. I also might return `undefined` instead of raising my own `ActionError`. It seems a bit obtuse to require users to _always_ check that `data` exists for every call to an action.

Reply:
I didn't think about `data` having a type of `returnType | undefined` which makes me think differently about this. If you trust your APIs then this becomes a bit of a burden. In which case I don't mind the `.safe()` as much.